gril: follow struct size change in data call list

peat-psuwit opened the pull request:
There are changes in RIL's data call list response in version 9 (or 10)
and 11 of the response. This cause a problem when there are more than 1
entry in the list.

This commit checks the version field to find out if it needs to read the
extra field(s). This fixes the problem where the device can randomly
become unable to send or receive an MMS message.

Note that there's conflicting information between ril.h [1] and
RIL.java [2] about the availability of the pcscf field, which I choose
to follow ril.h.

[1] https://github.com/LineageOS/android_hardware_ril/blob/cm-14.1/include/telephony/ril.h#L489
[2] https://github.com/LineageOS/android_frameworks_opt_telephony/blob/cm-14.1/src/java/com/android/internal/telephony/RIL.java#L4024

UniversalSuperBox commented on gril/grilunsol.c:
```diff
@@ -345,6 +368,8 @@ struct ril_data_call_list *g_ril_unsol_parse_data_call_list(GRil *gril,
 	g_free(raw_dns);
 	g_free(raw_gws);
+	if (raw_pcscf != NULL)
+		g_free(raw_pcscf);
```
On discussion with you, it appears raw_pcscf is not freed in anything but the error case.

peat-psuwit replied:
That should be fixed.

peat-psuwit committed:
gril: follow struct size change in data call list

There are changes in RIL's data call list response in version 9 (or 10)
and 11 of the response. This cause a problem when there are more than 1
entry in the list.

This commit checks the version field to find out if it needs to read the
extra field(s). This fixes the problem where the device can randomly
become unable to send or receive an MMS message.

Note that there's conflicting information between ril.h [1] and
RIL.java [2] about the availability of the pcscf field, which I choose
to follow ril.h.

[1] https://github.com/LineageOS/android_hardware_ril/blob/cm-14.1/include/telephony/ril.h#L489
[2] https://github.com/LineageOS/android_frameworks_opt_telephony/blob/cm-14.1/src/java/com/android/internal/telephony/RIL.java#L4024

peat-psuwit commented:
I also move the malformed parcel check a bit below to prevent over-reading for one iteration from triggering double-free in the next iteration's error.

UniversalSuperBox reviewed (approved):
Excellent, good work.
